### PR TITLE
Port micro operator EXP and its test code

### PR DESCRIPTION
**Pull request by rsun-bdti**
Combined PR4 and PR5 for issue #45415. This PR, if merged, should finish the issue.

### CODE REVIEW

File: tensorflow/lite/micro/kernels/exp.cc [hunk 1]
```diff
@@ -12,78 +12,69 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-#include "tensorflow/lite/kernels/internal/tensor.h"
 #include "tensorflow/lite/kernels/internal/tensor_ctypes.h"
 #include "tensorflow/lite/kernels/kernel_util.h"
+#include "tensorflow/lite/micro/kernels/kernel_util.h"
 
 namespace tflite {
 namespace ops {
-namespace builtin {
+namespace micro {
 namespace exp {
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/exp.cc:26]: we're going with a flat namespace in micro.

See https://github.com/tensorflow/tensorflow/commit/b13e49836a92241d653e96225552bd8df3bfd92e and https://github.com/tensorflow/tensorflow/blob/b0fcce8080f94b44195c429929361a420e9255cc/tensorflow/lite/micro/kernels/conv.cc

Tagging @rkuester to make sure that this is part of the updated OP porting guide.

File: tensorflow/lite/micro/kernels/exp.cc [hunk 2]
```diff
@@ -12,78 +12,69 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-  ExpContext op_context(context, node);
-
-#define TF_LITE_EXP(kernel_type, data_type)                               \
-  kernel_type::Exp<data_type>(GetTensorData<data_type>(op_context.input), \
-                              NumElements(op_context.input),              \
-                              GetTensorData<data_type>(op_context.output))
+  const TfLiteEvalTensor* input =
+      tflite::micro::GetEvalInput(context, node, 0);
+  TfLiteEvalTensor* output =
+      tflite::micro::GetEvalOutput(context, node, 0);
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/exp.cc:49]: pull the magic constants out to a constexpr.

https://github.com/tensorflow/tensorflow/blob/b0fcce8080f94b44195c429929361a420e9255cc/tensorflow/lite/micro/kernels/conv.cc#L31-L34

File: tensorflow/lite/micro/kernels/micro_ops.h [hunk 3]
```diff
@@ -54,6 +54,7 @@ TfLiteRegistration Register_CONCATENATION();
 TfLiteRegistration Register_COS();
 TfLiteRegistration Register_DEQUANTIZE();
 TfLiteRegistration Register_EQUAL();
+TfLiteRegistration Register_EXP();
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/micro_ops.h:57]: move to the tflite namespace section up top.

**Review by advaitjain — CHANGES_REQUESTED** [review]

**Comment by advaitjain** [comment]
also, for clang-format errors you can use format_code.py:
https://github.com/tensorflow/tensorflow/blob/b0fcce8080f94b44195c429929361a420e9255cc/tensorflow/lite/micro/tools/ci_build/test_code_style.sh#L77-L84

and pass in --fix for an in-place fix.

You will need a very recent version of clang-format. This is what we use as part of CI:
https://github.com/tensorflow/tensorflow/blob/b0fcce8080f94b44195c429929361a420e9255cc/tensorflow/tools/ci_build/Dockerfile.micro#L8-L9


Ideally we would install git commit hooks, but we need a bit more setup for that.

File: tensorflow/lite/micro/kernels/exp.cc (hunk 1 above)
Comment (rkuester) [reply tensorflow/lite/micro/kernels/exp.cc:26]: Indeed [it is](https://github.com/rkuester/tensorflow/blame/e6010bf77a4b9341f1ddd7fe85d6e07a3d75e843/tensorflow/lite/micro/kernels/PORTING.md#L60).

Comment (rsun-bdti) [reply tensorflow/lite/micro/kernels/exp.cc]: changed to flat namespace.

File: tensorflow/lite/micro/kernels/exp.cc (hunk 2 above)
Comment (rsun-bdti) [reply tensorflow/lite/micro/kernels/exp.cc]: fixed.

File: tensorflow/lite/micro/kernels/micro_ops.h (hunk 3 above)
Comment (rsun-bdti) [reply tensorflow/lite/micro/kernels/micro_ops.h]: fixed.

**Comment by rsun-bdti** [comment]
@advaitjain : thanks for the detailed review & feedback. I have addressed all issues but one. Still working on the clang-format errors.

**Comment by rsun-bdti** [comment]
@advaitjain clang-format errors have been fixed; all checks have passed

**Review by advaitjain — APPROVED** [review]

**Comment by advaitjain** [comment]
copybara failed, going to trigger again.

**Review by gbaned — APPROVED** [review]

**Pull request merged**